Fix bug for SimpliSafe 2 systems repeatedly saying "your settings have been synchronised"

nzapponi opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Bug fix caused by #41080.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #42192
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15356

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

nzapponi committed:
Fix bug for SimpliSafe 2 systems

nzapponi commented:
Note that I haven't been able to test this in the real world as I do not own a SimpliSafe gen 2.

bachya commented on homeassistant/components/simplisafe/__init__.py:
```diff
@@ -500,7 +500,7 @@ async def async_update(self):
         async def async_update_system(system):
             """Update a system."""
-            await system.update(cached=False)
+            await system.update(cached=system.version != 3)
```
This is messy. Just return the update if the system version isn't 3.

nzapponi replied:
The reason why I did it this way is because the `await system.update()` call already existed before I added sensor support, and I thought there would be a reason for it.

What I changed vs the original is the argument `cached=False`.

If you don't think there's any reasons to return something on `async_update_system(system)` for SimpliSafe 2 systems, then yes, I can update it following your suggestion.

What do you think @bachya ?

bachya replied:
I see; thanks for that clarification. In that case, let's stick with what you have.

bachya commented on homeassistant/components/simplisafe/binary_sensor.py:
```diff
@@ -51,11 +51,12 @@ async def async_setup_entry(hass, entry, async_add_entities):
-                sensors.append(SimpliSafeBinarySensor(simplisafe, system, sensor))
-            if sensor.type in SUPPORTED_BATTERY_SENSOR_TYPES:
-                sensors.append(SimpliSafeSensorBattery(simplisafe, system, sensor))
+        if system.version == 3:
```
Why are we checking the system version every time through the loop? I would check for system version before this loop and if version 2 is found, log a message (`INFO`-level) and return without doing anything else.

nzapponi replied:
I'm doing it in the loop because I assume a customer may have a SimpliSafe 2 system *and* a SimpliSafe 3 system in one account, meaning that one would work and the other wouldn't. Do you agree @bachya ?

bachya replied:
Apologies, my language was unclear – we still need a loop, but we can short-circuit it sooner for V2 systems:

```python
async def async_setup_entry(hass, entry, async_add_entities):
"""Set up SimpliSafe binary sensors based on a config entry."""
    simplisafe = hass.data[DOMAIN][DATA_CLIENT][entry.entry_id]
    sensors = []

    for system in simplisafe.systems.values():
        if system.version == 2:
            _LOGGER.info("Skipping sensor setup for V2 system: %s", system.system_id)
            continue

        for sensor in system.sensors.values():
            if sensor.type in SUPPORTED_SENSOR_TYPES:
                sensors.append(SimpliSafeBinarySensor(simplisafe, system, sensor))
            if sensor.type in SUPPORTED_BATTERY_SENSOR_TYPES:
                sensors.append(SimpliSafeSensorBattery(simplisafe, system, sensor))

    async_add_entities(sensors)
```

bachya commented on homeassistant/components/simplisafe/sensor.py:
```diff
@@ -16,6 +16,7 @@ async def async_setup_entry(hass, entry, async_add_entities):
         [
             SimplisafeFreezeSensor(simplisafe, system, sensor)
             for system in simplisafe.systems.values()
+            if system.version == 3
```
Same as above.

nzapponi replied:
Same as above.

bachya replied:
```python
async def async_setup_entry(hass, entry, async_add_entities):
    """Set up SimpliSafe freeze sensors based on a config entry."""
    simplisafe = hass.data[DOMAIN][DATA_CLIENT][entry.entry_id]
    sensors = []

    for system in simplisafe.systems.values():
        if system.version == 2:
            _LOGGER.info("Skipping sensor setup for V2 system: %s", system.system_id)
            continue

        for sensor in system.sensors.values():
            if sensor.type == EntityTypes.temperature:
                sensors.append(SimplisafeFreezeSensor(simplisafe, system, sensor))

    async_add_entities(sensors)
```

bachya reviewed (changes requested):
This also needs to be accompanied by a documentation PR that explains (a) the 30-second update interval (to ward off inevitable future "issues" around it) and (b) that sensors are only monitored on V3 systems.

nzapponi committed:
Improved loop and logging for SimpliSafe 2

nzapponi commented:
Added docs @bachya , see https://github.com/home-assistant/home-assistant.io/pull/15356
30 seconds was already in there as per previous PR. I added a comment on SimpliSafe V2.

bachya reviewed (approved):
Thank you!
